File: Editorial/3418_Maximum_Amount_of_Money_Robot_Can_Earn/Python3/1_Memoization_Search.py
class Solution:
    def maximumAmount(self, coins: List[List[int]]) -> int:
        m, n = len(coins), len(coins[0])
        memo = [[[-inf] * 3 for _ in range(n)] for _ in range(m)]

        def dfs(i: int, j: int, k: int) -> int:
            if i >= m or j >= n:
                return -inf

            x = coins[i][j]
            # arrive at the destination
            if i == m - 1 and j == n - 1:
                return max(0, x) if k > 0 else x

            if memo[i][j][k] != -inf:
                return memo[i][j][k]

            # not neutralize
            res = max(dfs(i + 1, j, k), dfs(i, j + 1, k)) + x
            if k > 0 and x < 0:
                # neutralize
                res = max(res, dfs(i + 1, j, k - 1), dfs(i, j + 1, k - 1))

            memo[i][j][k] = res
            return res

        return dfs(0, 0, 2)

File: Editorial/3501-3600/3573_Best_Time_to_Buy_and_Sell_Stock_V/TypeScript/1_Memoization_Search.ts
function maximumProfit(prices: number[], k: number): number {
    const n = prices.length;
    const memo: number[][][] = Array(n)
        .fill(null)
        .map(() =>
            Array(k + 1)
                .fill(null)
                .map(() => Array(3).fill(-1)),
        );

    const dfs = (i: number, j: number, state: number): number => {
        if (j == 0) {
            return 0;
        }
        if (i == 0) {
            return state == 0 ? 0 : state == 1 ? -prices[0] : prices[0];
        }
        if (memo[i][j][state] !== -1) {
            return memo[i][j][state];
        }
        const p = prices[i];
        let res: number;
        if (state === 0) {
            res = Math.max(
                dfs(i - 1, j, 0),
                Math.max(dfs(i - 1, j, 1) + p, dfs(i - 1, j, 2) - p),
            );
        } else if (state === 1) {
            res = Math.max(dfs(i - 1, j, 1), dfs(i - 1, j - 1, 0) - p);
        } else {
            res = Math.max(dfs(i - 1, j, 2), dfs(i - 1, j - 1, 0) + p);
        }
        memo[i][j][state] = res;
        return res;
    };

    return dfs(n - 1, k, 0);
}
